docs/developers/pass_pipeline.md

```diff
@@ -0,0 +1,886 @@
+      "real_dynamic_slice",
+      [NoSideEffect, AllElementTypesMatch<["operand", "result"]>,
+       AllTypesMatch<["start_indices", "limit_indices", "strides"]>]> {
+  let summary = "Real Dynamic Slice operator";
+  let description = [{
+    The dynamic shape version of SliceOp. Extracts a sub-array from the input
+    array according to start_indices, limit_indices and strides. Expect
+    start_indices/limit_indices/strides to be statically shaped and matching
+    the rank of the input.
+  }];
+  let arguments = (ins
+    HLO_Tensor:$operand,
+    HLO_DimensionTensor:$start_indices,
+    HLO_DimensionTensor:$limit_indices,
+    HLO_DimensionTensor:$strides
+  );
+  let results = (outs HLO_Tensor:$result);
+  let hasCanonicalizer = 1;
+  let hasCustomHLOConverter = 1;
+}
+```
+
+As you may have noticed, the major difference is that, in the static shape
+version the start/limit/strides are attributes, which means that the subgraph
+that produce them must be constant folded at compile time given each set of
+input shapes. While, for the dynamic shape version, these are regular inputs,
+which means the IR is able to express a frontend compute graph in dynamic shape
+semantics.
+
+* Shape calculation, placer, buffer management and runtime support
+
+One challenge of dynamic shape compiler is that, compiling is a static action,
+while we aim to handle dynamic behavior. In order to fully support dynamic
+shapes, the compiler must be able to do adaptive shape inference and generate
+code not only for data calculation, but also for shape calculation. The
+calculated shapes of each buffer will guide the buffer allocation and launch
+dimension at runtime. BladeDISC does bufferization during the conversion from
+mhlo to lmhlo. Rather than using an interpreter or virtual machine, BladeDISC
+compiles and generates the code of computations on both host and device side,
+and also runtime flows (buffer management, kernel launch, et.al.). 
+
+Runtime Abstraction Layer (RAL) is brought in to isolate the complexity between
+compile time and different runtime environment, so that the compiler does not
+see the difference of a libtorch runtime or a tensorflow runtime, also the
+compiler will not bother with the management of stateful informations at
+runtime. Refer to [Runtime Abstraction Layer](https://tba) for more
+informations.
+
+* Performance
+
+The challenge of an AI compiler on performance is enlarged when the shapes are
+unknown or partially unknown. For the current version, BladeDISC makes an
+artificial seperation on compute-intensive part (gemm & convolution) and
+memory-intensive part (the rest calculations in general). For compute-intensive
+ops, different shapes may require different optimization to achieve the best
+performance. In order to balance the dynamism and performance, we implement an
+interface to choose the best kernel from a library according to different
+runtime shapes, and we do code generation for the memory intensive part. The
```

Comment:
Better to be consistent with either "memory-intensive" or "memory intensive" expression.